perception/map_based_prediction/src/map_based_prediction_node.cpp

```diff
@@ -0,0 +1,997 @@
+
+  return current_pose;
+}
+
+double MapBasedPredictionNode::calcRightLateralOffset(
+  const lanelet::ConstLineString2d & boundary_line, const geometry_msgs::msg::Pose & search_pose)
+{
+  std::vector<geometry_msgs::msg::Point> boundary_path(boundary_line.size());
+  for (size_t i = 0; i < boundary_path.size(); ++i) {
+    const double x = boundary_line[i].x();
+    const double y = boundary_line[i].y();
+    boundary_path[i] = tier4_autoware_utils::createPoint(x, y, 0.0);
+  }
+
+  return std::fabs(tier4_autoware_utils::calcLateralOffset(boundary_path, search_pose.position));
+}
+
+double MapBasedPredictionNode::calcLeftLateralOffset(
+  const lanelet::ConstLineString2d & boundary_line, const geometry_msgs::msg::Pose & search_pose)
+{
+  return -calcRightLateralOffset(boundary_line, search_pose);
+}
+
+void MapBasedPredictionNode::updateFuturePossibleLanelets(
+  const TrackedObject & object, const lanelet::routing::LaneletPaths & paths)
+{
+  std::string object_id = toHexString(object.object_id);
+  if (objects_history_.count(object_id) == 0) {
+    return;
+  }
+
+  std::vector<lanelet::ConstLanelet> & possible_lanelets =
+    objects_history_.at(object_id).back().future_possible_lanelets;
+  for (const auto & path : paths) {
+    for (const auto & lanelet : path) {
+      bool not_in_buffer = std::find(possible_lanelets.begin(), possible_lanelets.end(), lanelet) ==
+                           possible_lanelets.end();
+      if (not_in_buffer) {
+        possible_lanelets.push_back(lanelet);
+      }
+    }
+  }
+}
+
+void MapBasedPredictionNode::addReferencePaths(
+  const TrackedObject & object, const lanelet::routing::LaneletPaths & candidate_paths,
+  const float path_probability, const ManeuverProbability & maneuver_probability,
+  const Maneuver & maneuver, std::vector<PredictedRefPath> & reference_paths)
+{
+  if (!candidate_paths.empty()) {
+    updateFuturePossibleLanelets(object, candidate_paths);
+    const auto converted_paths = convertPathType(candidate_paths);
+    for (const auto & converted_path : converted_paths) {
+      PredictedRefPath predicted_path;
+      predicted_path.probability = maneuver_probability.at(maneuver) * path_probability;
+      predicted_path.path = converted_path;
+      predicted_path.maneuver = maneuver;
+      reference_paths.push_back(predicted_path);
+    }
+  }
+}
+
+ManeuverProbability MapBasedPredictionNode::calculateManeuverProbability(
+  const Maneuver & predicted_maneuver, const lanelet::routing::LaneletPaths & left_paths,
+  const lanelet::routing::LaneletPaths & right_paths,
+  const lanelet::routing::LaneletPaths & center_paths)
+{
+  float left_lane_change_probability = 0.0;
+  float right_lane_change_probability = 0.0;
+  float lane_follow_probability = 0.0;
+  if (!left_paths.empty() && predicted_maneuver == Maneuver::LEFT_LANE_CHANGE) {
+    left_lane_change_probability = 1.0;
+    right_lane_change_probability = 0.0;
+    lane_follow_probability = 0.9;
```

Comment:
avoid direct insertion. Just define `constexpr float LF_PROB_WHEN_LC = 0.9` in this function is enough for now.

Reply:
done